tests/BookReader/BookModel.test.js

```diff
@@ -0,0 +1,175 @@
+            {width: 100, height: 2300},
+            {width: 400, height: 2400},
+        ];
+        const data = deepCopy(SAMPLE_DATA);
+        Object.assign(data[0][0], sizes[0]);
+        Object.assign(data[1][0], sizes[1]);
+        Object.assign(data[1][1], sizes[2]);
+        Object.assign(data[2][0], sizes[3]);
+        const bm = new BookModel({ data });
+        expect(bm.getMedianPageSize()).toEqual({ width: 300, height: 2300 });
+    });
+
+    test('caches result', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        const firstResult = bm.getMedianPageSize();
+        expect(bm.getMedianPageSize()).toBe(firstResult);
+        expect(bm.getMedianPageSize()).toBe(firstResult);
+        expect(bm.getMedianPageSize()).toBe(firstResult);
+    });
+});
+
+describe('getPageIndices', () => {
+    test('handles unique indices', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        expect(bm.getPageIndices('1')).toEqual([0]);
+    });
+    
+    test('handles invalid indices', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        expect(bm.getPageIndices('foo')).toEqual([]);
+        expect(bm.getPageIndices('123')).toEqual([]);
+    });
+
+    test('handles non-unique indices', () => {
+        const data = deepCopy(SAMPLE_DATA);
+        data[0][0].pageNum = '2';
+        const bm = new BookModel({ data });
+        expect(bm.getPageIndices('2')).toEqual([0, 1]);
+    });
+
+    test('handles n-prefixed pages', () => {
+        const data = deepCopy(SAMPLE_DATA);
+        data[0][0].pageNum = 'n0';
+        const bm = new BookModel({ data });
+        expect(bm.getPageIndices('n0')).toEqual([0]);
+    });
+});
+
+test('getPageName', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPageName(0)).toEqual('Page 1');
+    expect(bm.getPageName(1)).toEqual('Page 2');
+});
+
+describe('numLeafs', () => {
+    test('reads from data', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA });
+        expect(bm.getNumLeafs()).toBe(4);
+    });
+    
+    test('numLeafs can be overriden', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA, numLeafs: 12 });
+        expect(bm.getNumLeafs()).toBe(12);
+    });
+});
+
+test('getPageNum', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPageNum(0)).toBe('1');
+});
+
+test('getPageProp gets values from correct document', () => {
+    const bm = new BookModel({ data: SAMPLE_DATA });
+    expect(bm.getPageProp(1, 'uri')).toBe(SAMPLE_DATA[1][0].uri);
+});
+
+describe('getSpreadIndices', () => {
+    test('Works for left to right text', () => {
+        const bm = new BookModel({ data: SAMPLE_DATA, pageProgression: 'lr' });
+        expect(bm.getSpreadIndices(1)).toEqual([1, 2]);
+        expect(bm.getSpreadIndices(2)).toEqual([1, 2]);
+    });
+});
+
+describe('leafNumToIndex', () => {
+    test('returns leafNum as is if no leafNum in data', () => {
```

Comment:
suggestion: change description to `returns index as when there isn't a leafNum in data`

Reply:
Changed to `returns number inputs if no leafNum in data`